transport/transporters/s3/transporter/transporter.go

```diff
@@ -0,0 +1,322 @@
+		txnsWritten,
+		statsChan,
+		log,
+		client,
+		bucketName,
+		keySpace,
+	}
+}
+
+// NewTransporter returns a s3 transporter
+func NewTransporter(shutdownHandler shutdown.ShutdownHandler,
+	inputChan <-chan transport.Batch,
+	txnsWritten chan<- *ordered_map.OrderedMap,
+	statsChan chan stats.Stat,
+	log logrus.Entry,
+	id int,
+	bucketName string,
+	keySpace string,
+	awsRegion *string,
+	awsAccessKeyId *string,
+	awsSecretAccessKey *string,
+	endpoint *string,) transport.Transporter {
+
+	var sess *session.Session
+
+	creds := credentials.NewStaticCredentials(*awsAccessKeyId, *awsSecretAccessKey, "")
+
+	if *endpoint != "" {
+		sess = session.Must(session.NewSession(&aws.Config{
+			Region:      		aws.String(*awsRegion),
+			Credentials: 		creds,
+			S3ForcePathStyle: 	aws.Bool(true),
+			Endpoint:    		aws.String(*endpoint),
+		}))
+	} else {
+		sess = session.Must(session.NewSession(&aws.Config{
+			Region:      aws.String(*awsRegion),
+			Credentials: creds,
+		}))
+	}
+
+	client := s3.New(sess)
+
+	return NewTransporterWithInterface(
+		shutdownHandler,
+		inputChan,
+		txnsWritten,
+		statsChan,
+		log,
+		id,
+		bucketName,
+		keySpace,
+		client)
+}
+
+// shutdown idempotently closes the output channel
+func (t *S3Transporter) shutdown() {
+	t.log.Info("shutting down transporter")
+	t.shutdownHandler.CancelFunc() // initiate shutdown on other modules as well
+
+	if r := recover(); r != nil {
+		t.log.Warnf("Recovered in S3Transporter %s", r)
+	}
+
+	defer func() {
+		// recover if channel is already closed
+		recover()
+	}()
+
+	t.log.Debug("closing progress channel")
+	close(t.txnsWritten)
+}
+
+
+// transport does a PUT on a full batch as a single key/file to S3
+func (t *S3Transporter) transport(ctx context.Context, messagesSlice []*marshaller.MarshalledMessage) (error, bool) {
+	var cancelled bool
+
+	var buf bytes.Buffer
+	gz := gzip.NewWriter(&buf)
+
+	select {
+	case <-ctx.Done():
+		t.log.Debug("received terminateCtx cancellation")
+		return nil, true
+	default:
+	}
+
+	// add all messages into a gzipped buffer
+	for _, msg := range messagesSlice {
+		if _, err := gz.Write(msg.Json); err != nil {
+			return err, cancelled
+		}
+		gz.Write([]byte("\n"))
+	}
+
+	if err := gz.Flush(); err != nil {
+		return err, cancelled
+	}
+	if err := gz.Close(); err != nil {
+		return err, cancelled
+	}
+
+	gz.Reset(ioutil.Discard)
+
+	firstWalStart := messagesSlice[0].WalStart
+
+	// Free up messagesSlice now that we're done with it
+	messagesSlice = nil
+
+	// Clear out every allocation in the conversion
+	byteArray := buf.Bytes()
+	buf.Reset()
+
+	byteReader := bytes.NewReader(byteArray)
+	byteArray = nil
+
+	// Partition the S3 keys into days
+	year, month, day, full := TimeSource.DateString()
+	baseFilename := fmt.Sprintf("%s-%d", full, firstWalStart)
+
+	fullKey := key_join(true, t.keySpace, year, month, day, baseFilename)
+
+	// Do the upload and let S3 handle retries
+	_, err := t.client.PutObjectWithContext(ctx, &s3.PutObjectInput{
```

Comment:
How do retries work with the S3 client? Are the errors we get back from the call all terminal?

Reply:
It uses S3 client based retries with their own retry logic:

https://github.com/aws/aws-sdk-go/blob/master/aws/config.go#L92
https://github.com/aws/aws-sdk-go/blob/master/aws/client/default_retryer.go#L34

As far as I can tell, you can't get a return of a count of how many times it retried and then succeeded. Once it exhausts it's retries then it fails completely.